packages/kbn-esql-validation-autocomplete/src/validation/validation.ts

```diff
@@ -432,56 +430,41 @@ function validateFunction(
-        const everyArgInListMessages = outerArg
-          .map((arg) => {
-            return [
-              validateFunctionLiteralArg,
-              validateNestedFunctionArg,
-              validateFunctionColumnArg,
-            ].flatMap((validateFn) => {
-              return validateFn(
-                astFunction,
-                arg,
-                { ...argDef, type: extractedType },
-                references,
-                parentCommand
-              );
-            });
-          })
-          .filter((ms) => ms.length);
-        if (everyArgInListMessages.length) {
-          failingSignature.push(
-            getMessageFromId({
-              messageId: 'wrongArgumentType',
-              values: {
-                name: astFunction.name,
-                argType: argDef.type,
-                value: `(${getAllArrayValues(outerArg).join(', ')})`,
-                givenType: `(${getAllArrayTypes(outerArg, parentCommand, references).join(', ')})`,
-              },
-              locations: {
-                min: (outerArg[0] as ESQLSingleAstItem).location.min,
-                max: (outerArg[outerArg.length - 1] as ESQLSingleAstItem).location.max,
-              },
-            })
-          );
-        }
-        return;
-      }
-      const wrappedArg = Array.isArray(outerArg) ? outerArg : [outerArg];
-      for (const actualArg of wrappedArg) {
-        const argValidationMessages = [
-          validateFunctionLiteralArg,
-          validateNestedFunctionArg,
-          validateFunctionColumnArg,
-        ].flatMap((validateFn) => {
-          return validateFn(astFunction, actualArg, argDef, references, parentCommand);
-        });
-        failingSignature.push(...argValidationMessages);
+
+      // check every element of the argument (may be an array of elements, or may be a single element)
+      const hasMultipleElements = Array.isArray(outerArg);
+      const argElements = hasMultipleElements ? outerArg : [outerArg];
+      const singularType = extractSingularType(argDef.type);
+      let messagesFromAllArgElements = argElements
+        .map((arg) => {
+          return [
+            validateFunctionLiteralArg,
+            validateNestedFunctionArg,
+            validateFunctionColumnArg,
+          ].flatMap((validateFn) => {
+            return validateFn(
+              astFunction,
+              arg,
+              { ...argDef, type: singularType },
+              references,
+              parentCommand
+            );
+          });
+        })
+        .flat();
+
+      if (isArrayType(argDef.type) && hasMultipleElements) {
+        messagesFromAllArgElements = collapseWrongArgumentTypeMessages(
+          messagesFromAllArgElements,
+          outerArg,
+          astFunction.name,
+          argDef.type,
+          parentCommand,
+          references
+        );
       }
+
+      failingSignature.push(...messagesFromAllArgElements);
```

Comment:
This variable re-assignment is easy to be missed while reading the code.
At this point I think a ternary operation would be easier to read:

```suggestion
      const shouldCollapseMessages = isArrayType(argDef.type) && hasMultipleElements;
      failingSignature.push(...(
        shouldCollapseMessages ? 
          messagesFromAllArgElements : 
          collapseWrongArgumentTypeMessages(
            messagesFromAllArgElements,
            outerArg,
            astFunction.name,
            argDef.type,
            parentCommand,
            references
          )
        )
      );
```